[W.I.P] Proposal: Pluggable SBOM Generation

Dentrax opened the pull request:
Fixes https://github.com/goharbor/harbor/issues/16397

Signed-off-by: Furkan <furkan.turkal@trendyol.com>
Co-authored-by: Batuhan <batuhan.apaydin@trendyol.com>

cc @hectorj2f @luhring @wagoodman @developer-guy

Dentrax commented:
Thanks for feedbacks, @itaysk! We updated the proposal to make more generic SBOM generation plugin system. Eventually, other friends can easily implement their own SBOM generation methods.

Vad1mo commented:
I support having an SBOM. What about displaying the SBOM content in Harbor? Would that not also make sense?

hectorj2f commented on proposals/new/sbom-integration.md:
```diff
@@ -0,0 +1,86 @@
+## Goal
+
+* Create a generic SBOM generation plugin system
+* Support any SBOM generation tool
```
Which format would be supported as output SPDX, CycloneDX or both ?

Dentrax replied:
Thanks, why not both!

hectorj2f commented on proposals/new/sbom-integration.md:
```diff
@@ -0,0 +1,86 @@
+
+* Create a generic SBOM generation plugin system
+* Support any SBOM generation tool
+* Create signed SBOM attestations
```
Do you refer to creating attestations with the content of the SBOM as explained here https://github.com/sigstore/cosign/issues/1742 ?

Dentrax replied:
Exactly, referred this proposal.

Dentrax commented:
> I support having an SBOM. What about displaying the SBOM content in Harbor? Would that not also make sense?

We definitely should! Currently this proposal lack of some details and it's not deeply-technical as [Pluggable Image Vulnerability Scanning](https://github.com/goharbor/community/blob/main/proposals/pluggable-image-vulnerability-scanning_proposal.md) proposal. So let me mark this as W.I.P.

Since we dont know much about Harbor’s overall domain and internals, looking forward to any contribution from the community!

danielpacak commented on proposals/new/sbom-integration.md:
```diff
@@ -0,0 +1,86 @@
+
+## Proposal
+
+Create a generic SBOM generation plugin system in Harbor for Software Bill of Materials (SBOM) from container images and filesystems. Since Cosign already supported by Harbor, we can store signatures in an OCI registry next to the container image, and can be located via a simple name scheme. The Cosign spec allows SBOM information to be embedded into the cosign artifact. [^2]
```
Not sure what does it mean "generic plugin system" but have you considered defining MimeType and report format and reuse [Pluggable Scanners API](https://github.com/goharbor/pluggable-scanner-spec)? We've been discussing SBOB and licence scanners in the context of pluggable scanners from the get go and it should be easy to leverage it. See also existing issue that was created in the past to do exactly what you requested https://github.com/goharbor/pluggable-scanner-spec/issues/10 /cc @steven-zou

steven-zou replied:
For security scanning, we should follow the [Pluggable Scanners API](https://github.com/goharbor/pluggable-scanner-spec).

As Harbor has introduced accessories for OCI artifacts, an improvement I can see now is to enhance Harbor to support storing any scanning results as an accessory of OCI artifacts. cc @wy65701436

OrlinVasilev commented:
@Dentrax @danielpacak do we have anything new on this one ?

Dentrax commented:
> Do we have anything new on this one ?

Sorry for the long delay; we dropped the ball here.

I couldn't find much free time to get into it, but we should do some fixes/changes according to reviews.

Feel free to commit this branch!

/cc @developer-guy

Vad1mo commented:
Would love to see SBOM list in Harbor, @Dentrax Can you address the questions from the community?

Dentrax commented:
> Would love to see SBOM list in Harbor, @Dentrax Can you address the questions from the community?

Hey! Thanks for reminding, almost forgot this one! I just addressed the reviews.

Since I'm not so familiar with Harbor internals, I'm not so sure what I can include on this doc additionally. Feel free to take over/carry/contribute this one if you want to include low-level system/arch designs, or the spec. Currently, it's like a very high-level of "why we should add SBOM support in Harbor" rather than "how" - that's the best thing I can do as an end-user, unfortunately.

---

@hectorj2f Thanks for reviews!

> What about displaying the SBOM content in Harbor? Would that not also make sense?

@Vad1mo Cool idea! I added this idea on the doc. But couldn't go into details much, It'd be nice to show vulnerability detail on the UI about each dep, if possible.

cc @OrlinVasilev @danielpacak
